i18n-helpers/src/lib.rs

```diff
@@ -54,33 +56,215 @@ pub fn extract_paragraphs(text: &str) -> impl Iterator<Item = (usize, &str)> {
+    }
+
+    /// Push a new Markdown event into the accumulator.
+    fn push_event(&mut self, evt: Event, offset: Range<usize>) {
+        dbg!(&offset);
+        match dbg!(evt) {
+            Event::Start(Tag::Emphasis | Tag::Strong | Tag::Strikethrough) => {
+                self.push_message(Message::Text(offset))
+            }
+            Event::Start(Tag::CodeBlock(_)) => {
+                // note that this offset contains the entire codeblock!
+                self.push_message(Message::Code(offset));
+                self.in_codeblock = true;
+            }
+            Event::Start(_) => self.finish_message(),
+
+            Event::End(Tag::Emphasis | Tag::Strong | Tag::Strikethrough) => {
+                self.push_message(Message::Text(offset))
+            }
+            Event::End(Tag::CodeBlock(_)) => {
+                self.in_codeblock = false; // (assumes codeblocks aren't nested!)
+                self.finish_message();
+            }
+            Event::End(_) => self.finish_message(),
+
+            Event::Text(_) if !self.in_codeblock => {
+                self.push_message(Message::Text(offset))
+            }
+            Event::Code(content) => self.push_message(Message::Code(offset)),
+            Event::Html(content) => self.push_message(Message::Html(offset)),
+            // TODO: list items
+            // TODO: headings
+            _ => {}
+        }
+    }
+
+    /// Get the resulting list of messages.
+    fn into_msgs(mut self) -> Vec<(usize, &'a str)> {
+        let parser = Parser::new(self.text);
+        for (evt, offset) in parser.into_offset_iter() {
+            self.push_event(evt, offset);
+        }
+        self.msgs
+            .iter()
+            .map(|msg| msg.as_range())
+            .map(|range| {
+                (
+                    self.line_number(range.start),
+                    &self.text[range.start..range.end],
+                )
+            })
+            .collect()
+    }
+}
+
+/// Extract translatable messages from the markdown text.
+///
+/// Returns a vector of (line number, text), where line numbers begin
+/// at 1.
+pub fn extract_msgs(text: &str) -> Vec<(usize, &str)> {
+    MsgAccumulator::new(text).into_msgs()
+}
+
 #[cfg(test)]
 mod tests {
     use super::*;
 
-    macro_rules! assert_iter_eq {
-        ($left_iter:expr, $right:expr) => {
-            assert_eq!($left_iter.collect::<Vec<_>>(), $right)
-        };
+    #[test]
+    fn offset_to_line_empty() {
+        assert_eq!(MsgAccumulator::new("").line_number(0), 1);
     }
 
     #[test]
-    fn test_extract_paragraphs_empty() {
-        assert_iter_eq!(extract_paragraphs(""), vec![]);
+    fn offset_to_line_multiline() {
+        let input = "abc\ndef\nghi";
+        let acc = MsgAccumulator::new(input);
+        let line_nums: Vec<_> = (0..=input.len()).map(|o| acc.line_number(o)).collect();
+
+        assert_eq!(
+            line_nums,
+            vec![
+                1, // a
```

Comment:
I think you could rewrite the test to something like

```
let char_line_nums = input.chars().enumerate().map(|(ch, idx)| (ch, acc.line_number(idx)).collect::<Vec<_>>();
```

Then your `assert_eq!` would not need the comments since the character is visible in the output.